### DISCUSSION TITLE
SSR Mode vue-cli/webpack

### DISCUSSION THREAD

**Discussion by artnova-david** [quasarframework/quasar#14169] [CLI - SSR mode]
Hi,
I'm currently working on an vue-cli/webpack app with Quasar integrated, this works fine in SPA mode.
I also successfully configured the app for SSR with vue-router and pinia, but I'm a bit stuck on Quasar giving me errors like
'window is not defined' or 'Cannot read properties of undefined (reading 'screen')'. I suppose this is because SSR-mode is not enabled for Quasar and therefore the server tries to execute components which are not available.

I know that the most optimal way to go is true quasar-cli, but this is not an option for this project. Is there a way to configure/activate SSR mode for quasar in an existing  vue-cli app? And if so, how can I accomplish this?
I would really appreciate any help.

Thanks in advance
Kind Regards

**Comment by TobyMosque** [comment]
I'm not sure if that will work, but u can try to set the flags:
`__QUASAR_SSR__`, `__QUASAR_SSR_SERVER__`, `__QUASAR_SSR_CLIENT__`

https://github.com/quasarframework/quasar/blob/dev/app-webpack/lib/quasar-conf-file.js#L316
https://github.com/quasarframework/quasar/blob/dev/app-webpack/lib/webpack/create-chain.js#L30

**Reply by artnova-david** [reply]
Hi TobyMosque,
Thanks for your answer.
I tried this, but unless I placed them in the wrong place (vue.config.js) this did not work.

```
config.
         plugin("feature-flags")
              .tap(([options]) => {
                   options.__QUASAR_SSR__ = JSON.stringify(true)
                    options.__QUASAR_SSR_SERVER__ = JSON.stringify(true)
               return [options]
           })
```

  I still get 'window is not defined' when starting express server.
  Kind Regards

**Reply by stefanvanherwijnen** [reply]
Correct me if I'm wrong, but I don't think Quasar ships a SSR bundle, i.e. all files in `dist` contain references to `window`.

You should import from `quasar/src`, but you will need to define the constants as pointed out by TobyMosque.

**Reply by TobyMosque** [reply]
IMHO: "but this is not an option for this project." is the same as "I see my boss/client as a monster, so I'm too afraid to discuss anything with him, so I will silently accept any requirement". 

Personally, i never encountered a boss/client who didn't accepted to modify a requirement, if the modification results in a smaller cost and a better code base.

**Reply by artnova-david** [reply]
@TobyMosque
There are many other reasons why "this is not an option for this project" could be possible, but "I see my boss/client as a monster, so I'm too afraid to discuss anything with him, so I will silently accept any requirement" is certainly is NOT one of them. I'm amazed you could derive al that information just from my question for help.

**Reply by TobyMosque** [reply]
I'm sorry for hear that, do you can enlist that list of valid technical reasons, this can be very useful to improve the Quasar CLI.

**Reply by artnova-david** [reply]
We integrated Quasar in an existing vue-cli app quite some time ago. All our build processes are configured around vue-cli. Quasar V1 made integration with vue-cli and webpack possible so we choose that direction at the time. Hence my question "Is there a way to configure/activate SSR mode for quasar in an existing vue-cli app? And if so, how can I accomplish this?"

**Reply by TobyMosque** [reply]
Migrate from the vue-cli to the quasar-cli isn't that hard, and adapt yout CICD scripts isn't that hard. i think that effort will pay off.
Doing the SSR thing by your self will only led you to a unknow area (not officially supported), your pay off will be the chance of appear random bugs on your app after each quasar update.

**Reply by artnova-david** [reply]
I understand that, and in the future migrating the app to Quasar-cli is a possibility, but for now and due to deadlines it is not an option.

Thanks for your help.
Kind regards?

**Reply by TobyMosque** [reply]
I undestand, IMHO you're taking more risks with your decision to not migrate.
as i said, you're alone here, and i really don't know how much time you'll need to figure out a workaround to your problem. and i don't need to say, workarounds are a risk by themself.
